src/core/ext/transport/chttp2/transport/bin_decoder.h

```diff
@@ -0,0 +1,48 @@
+/*
+ *
+ * Copyright 2016, Google Inc.
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met:
+ *
+ *     * Redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *     * Neither the name of Google Inc. nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+#ifndef GRPC_CORE_EXT_TRANSPORT_CHTTP2_TRANSPORT_BIN_DECODER_H
+#define GRPC_CORE_EXT_TRANSPORT_CHTTP2_TRANSPORT_BIN_DECODER_H
+
+#include <grpc/support/slice.h>
+
+/* base64 decode a slice with pad chars. Returns a new slice, does not take
+   ownership of the input */
+gpr_slice grpc_chttp2_base64_decode(gpr_slice input);
+
+/* base64 decode a slice without pad chars, data length is needed. Returns a new
+   slice, does not take ownership of the input */
+gpr_slice grpc_chttp2_base64_decode_with_length(gpr_slice input,
```

Comment:
That would be the idea. If you can think of a more efficient interface,
speak up :) I'm drawing inspiration from zlib here.

On Fri, May 13, 2016 at 10:52 AM Yuchen Zeng <notifications@github.com>
wrote:

> In src/core/ext/transport/chttp2/transport/bin_decoder.h
> <https://github.com/grpc/grpc/pull/6572#discussion_r63224118>:
>
> > + * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
> > + *
> > + */
> > +
> > +#ifndef GRPC_CORE_EXT_TRANSPORT_CHTTP2_TRANSPORT_BIN_DECODER_H
> > +#define GRPC_CORE_EXT_TRANSPORT_CHTTP2_TRANSPORT_BIN_DECODER_H
> > +
> > +#include <grpc/support/slice.h>
> > +
> > +/* base64 decode a slice with pad chars. Returns a new slice, does not take
> > +   ownership of the input */
> > +gpr_slice grpc_chttp2_base64_decode(gpr_slice input);
> > +
> > +/* base64 decode a slice without pad chars, data length is needed. Returns a new
> > +   slice, does not take ownership of the input */
> > +gpr_slice grpc_chttp2_base64_decode_with_length(gpr_slice input,
>
> Cool. So grpc_base64_decode_partial will decode the data until either
> input_end is reached or output_end is reached? It won't do any check on the
> length of input/output memory?
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly or view it on GitHub
> <https://github.com/grpc/grpc/pull/6572/files/d982bba41d8676f33431d540d8a4b0c15e426f1c#r63224118>
>